README.md


**Version 1.11.6** (20.06.2021)

## Configuration steps
* style.css - Rename Childthme
* Screenshot.png - Insert Logo
* configuration.php - Select needed custom classes in row number 81
* gulpfile.js - Insert development domain to use gulp watch in row number 26

### Style your theme with gulp - console actions
Go to the Childthme directory and run following actions and style your theme
* npm install
* npm install gulp
* gulp && gulp watch


CHANGELOG.md
# CHANGELOG

## 1.13.8 | 29.06.2022 <block options, demos>
* block style options
* javascript functions to reuse
* Demo for class with cpt

## 1.12.8 | 17.08.2021 <configurator>
* move configurator file to mastertheme and define class usage in function.php

## 1.11.8 | 17.08.2021 <customizer>
* add hamburger color to customizer
* mobile/desktop css class fix
* remove responsive file from frontend scss

## 1.11.7 | 15.07.2021 <deploy issue>
* css vars for black & white
* remove theme JS from vendors
* add default customizer css file
* remove vars and forms css file
* remove default styles for colors and fonts

## 1.11.6 | 20.06.2021 <input max-width, fieldset styles>
* fix input max width
* fieldset styles

## 1.11.5 | 11.06.2021 <default css to master>
* move default theme styles to master theme

## 1.11.5 | 03.06.2021 <sticky resize, centered blocks, header fix>
* check sticky header on resize
* add centered container for gutenberg group and cover block
* fix header responsive

## 1.11.4 | 27.03.2021 <gutenberg block styling, css clean, swiper>
* move spacing, block and search styles to structure
* improve swiper for multiple columns
* add css file for custom blocks

## 1.10.4 | 27.03.2021 <gutenberg blocks, sticky update>
* gutenberg container with full height
* gutenberg block style register container
* gutenberg block vimeo support from master theme
* sticky menu js script fix
* js function to detect if element is in viewport

## 1.9.4 | 08.03.2021 <gulp, js, mautic, anchor & inline fs>
* add gulp-babel to allow const and let
* add use strict to javascript
* add body and frontend class to additional JS css
* check element existens before setting event listener
* mautic class added to configuration selection
* IE javascript file for fixes
* Anchor position for sticky header // css solution
* Inline font size responsive solution
* JS PopUp container
* Swiper/Grid Gallery Gutenberg friendly

## 1.7.4 | 11.09.2020 <ajax safty, javascript 1.0 release>
* fix video max width
* ajax safty
* remove jQuery from base theme
* add cheatsheet for javascript

## 1.6.4 | 07.09.2020 <blog template style>
* block style on overview page

## 1.5.4 | 02.09.2020 <gutenberg blocks, header & footer>
* Demo for a Gutenberg block template
* Demo for a custom Gutenberg block pattern
* CSS for header & footer container with before and after values

## 1.4.4 | 14.08.2020 <backend title>
* Fix backend title font size override
* Backend post/page title small if title is hidden

## 1.4.3 | 13.08.2020 <master, body>
* Fix master theme name
* Set body margin and padding to 0

## 1.4.2 | 10.08.2020 <mobile menu>
* mobile menu scrollable

## 1.4.1 | 10.08.2020 <load sticky, css file structure>
* header sticky on page load
* custom styling files

## 1.3.1 | 29.07.2020 <fix classes register>
* register childtheme classes

## 1.3 | 29.07.2020 <register classes, backend css>
* register childtheme classes
* seperate general styles from backend

## 1.2 | 28.07.2020 <frontend/backend css>
* change default font
* move align wide & fullwidth to general
* input styles only for frontend
* styles also for backend

## 1.1 | 28.07.2020 <theme colors>
* change default colors

## 1.0 | 28.07.2020 <root files>
* childtheme requested files
* configuration files
* ajax file for future actions
* assets minification with gulp


assets/scripts/theme-base.js
/*!
 * All template base javascript functions
 *
 * @author      David Voglgsang
 * @version     1.2
 *
 */


/*==================================================================================
  SETTINGS
==================================================================================*/

'use strict';

/* Global values
/------------------------*/
const root = document.querySelector('html'),
      configuration = ajaxCall({action: 'configuration'}),
      language = root.getAttribute('lang'),
      isTouch = 'ontouchstart' in document.documentElement,
      body = root.querySelector('body'),
      header = body.querySelector('header'),
      mainMenu = header.querySelector('#menu-main-container'),
      hamburger = header.querySelector('.hamburger'),
      main = body.querySelector('main');

let position = root.scrollTop,
    scrollPosition = window.scrollY;


/* Touch Device
/------------------------*/
root.setAttribute('data-touch', isTouch);



/*==================================================================================
 BASE FUNCTIONS
==================================================================================*/

/* Debounce function
/------------------------*/
function debounce(func, wait, immediate) {
  var timeout;
  return function() {
    var context = this, args = arguments;
    var later = function() {
      timeout = null;
      if (!immediate) func.apply(context, args);
    };
    var callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) func.apply(context, args);
  };
};


/* Unique id generator
/------------------------*/
let uniqueID = () => {
  return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
}


/* Convert RGB color to HEX code
/------------------------*/
function rgb2hex(rgb) {
  rgb = rgb.match(/^rgb\((\d+),\s*(\d+),\s*(\d+)\)$/);
  function hex(x) {
    return ("0" + parseInt(x).toString(16)).slice(-2);
  }
  return "#" + hex(rgb[1]) + hex(rgb[2]) + hex(rgb[3]);
}


/* Convert HEX code to RGB color
/------------------------*/
function hex2rgb(hex, opacity) {
  hex = hex.replace('#', '');
  r = parseInt(hex.substring(0,2), 16);
  g = parseInt(hex.substring(2,4), 16);
  b = parseInt(hex.substring(4,6), 16);
  result = 'rgba(' + r + ', ' + g + ', ' + b +', ' + opacity/100 + ')';
  return result;
}


/* Add css rule to stylesheet
/------------------------*/
// Example: addCSSRule(document.styleSheets[0], "header", "float: left");
function addCSSRule(sheet, selector, rules, index) {
  if("insertRule" in sheet) {
    sheet.insertRule(selector + "{" + rules + "}", index);
  }
  else if("addRule" in sheet) {
    sheet.addRule(selector, rules, index);
  }
}

/* Slugify string
/------------------------*/
function slugify(Text){
  return Text.toLowerCase()
  .replace(/ /g,'-')
  .replace(/ä/g,'ae')
  .replace(/Ä/g,'ae')
  .replace(/ö/g,'oe')
  .replace(/Ö/g,'oe')
  .replace(/ü/g,'ue')
  .replace(/Ü/g,'ue')
  .replace(/ß/g,'ss')
  .replace(/[^\w-]+/g,'');
}


/* Check if element is visible
/------------------------*/
function isInViewport(el) {
  const rect = el.getBoundingClientRect();
  return (
    rect.top >= 0 &&
    rect.left >= 0 &&
    rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
    rect.right <= (window.innerWidth || document.documentElement.clientWidth)
  );
}


/* AJAX function
Example of calling ajax:
var configuration = {
  action: 'action to run after ajax request is done',
  path: 'file_path/filename.php',
  other_vars: 'vars content'
};
ajaxCall(configuration);
/------------------------*/
function ajaxCall(getdata) {
  // function file path
  if(getdata.path){
    var ajaxFile = "/" + getdata.path;
  } else {
    var ajaxFile = "/ajax.php";
  }
  // ajax is active - disable reload
  if(event){
    event.preventDefault();
  }
  // get data ready for request
  var access = 'access=granted&';
  var sendData = Object.keys(getdata).map(function (key) {
    return "" + key + "=" + getdata[key]; // line break for wrapping only
  }).join("&");
  // start request
  var request = new XMLHttpRequest();
  request.open('POST', theme_directory + ajaxFile, true);
  request.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
  request.onload = function() {
  if (this.status >= 200 && this.status < 400) {
      var results = JSON.parse(this.response);
      // console
      if(results.log){
        console.log(results.log);
      }
      // run function
      if(results.action){
        eval(results.action + '(' + this.response + ')');
      } else {
        // DEBUG: console.log("Action not defined in ajax function");
      }
    } else {
      // DEBUG: console.log("Ajax update failed");
    }
  };
  request.onerror = function() {
    // DEBUG: console.log("Ajax conection failed");
  };
  request.send(access + sendData);
}



/*==================================================================================
 THEME FUNCTIONS
==================================================================================*/

/* Settings from configuration file
/------------------------*/
function themeConfiguration(data){
  // check if file value exists and inline css are disabled
  if(data.content !== false && data.content.wp.HeaderCss !== "1"){
    // set color palette
    if(data.content.gutenberg && Array.isArray(data.content.gutenberg.ColorPalette)){
      // gutenberg is given and colors been added
      var colorPalette = data.content.gutenberg.ColorPalette;
      var keys = Object.keys(data.content.gutenberg.ColorPalette);
      for(var i=0; i<keys.length; i++){
        addCSSRule(document.styleSheets[0], 'body.frontend .has-' + slugify(colorPalette[i].key) + '-background-color', 'background-color: ' + colorPalette[i].value);
        addCSSRule(document.styleSheets[0], 'body.frontend .has-' + slugify(colorPalette[i].key) + '-color', 'color: ' + colorPalette[i].value);
      }
    }
  }
}


/* Sticky menu
/------------------------*/
function StickyMenu(action) {
  // get new scroll position and header height
  var scroll = window.scrollY,
      headerHeight = header.offsetHeight;
  if(action == "load" && body.classList.contains('stickyable') === false){
    // set main margin for sticky on load
    main.style.marginTop = headerHeight + 'px';
  } else {
    // check if header is stickyable
    if(body.classList.contains('stickyable') && scroll > scrollPosition){
      body.classList.add("sticky");
      // set main margin for sticky after scroll
      if(scroll > headerHeight || action == "load"){
        main.style.marginTop = headerHeight + 'px';
      }
    }
  }
  // reset sticky if sticky is not perminent
  if (scroll < headerHeight) {
    if(body.classList.contains('stickyable')){
      main.style.marginTop = '';
      body.classList.remove("sticky");
    }
  }
  // update main scroll position
  scrollPosition = scroll;
  // body css if page is not on top
  if(scrollPosition > 0){
    body.classList.add("scrolled");
  } else {
    body.classList.remove("scrolled");
  }
}
var debounceSticky = debounce(function() {
  StickyMenu();
}, 100);
var debounceStickyResize = debounce(function() {
  StickyMenu("load");
}, 100);
window.addEventListener('scroll', debounceSticky);
window.addEventListener('resize', debounceStickyResize);
window.onload = function() {
  StickyMenu("load");
};


/* Hamburger switch
/------------------------*/
function MenuToggler() {
  // show overlay
  mainMenu.classList.toggle("hidden_mobile");
  mainMenu.classList.toggle("hidden_desktop");
  // prevent content scrolling
  root.classList.toggle("noscroll");
  // identify active menu
  body.classList.toggle("active-menu");
}
if(hamburger) {
  hamburger.addEventListener('click', MenuToggler);
}


/* Action Links
/------------------------*/
// example: <span class="funcCall" data-ajax-action="true" data-action="DEMO" data-id="page">DEMO</span>
function funcCall(){
  // vars
  var get_ajax_action = this.getAttribute('data-ajax-action'),
      get_action = this.getAttribute('data-action'),
      get_id = this.getAttribute('data-id');
  // actions
  if(get_ajax_action){
    // run ajax function
    var config = {
      action: get_action,
      id: get_id
    };
    ajaxCall(config);
  } else if (get_action) {
    // run function
    eval(get_action + '()');
  }
}
var actionButtons = document.querySelectorAll('.funcCall');
if(actionButtons.length !== 0){
  Array.from(actionButtons).forEach(function(element) {
    element.addEventListener('click', funcCall);
    element.addEventListener('keypress', funcCall);
  });
}



/*==================================================================================
  THEME BLOCKS
==================================================================================*/

/* Toggle
/------------------------*/
function toggleBlock(){
  this.classList.toggle("active");
}
var toggleButtons = document.querySelectorAll('.toggle > .label');
if(toggleButtons.length !== 0){
  Array.from(toggleButtons).forEach(function(element) {
    element.addEventListener('click', toggleBlock);
    element.addEventListener('keypress', toggleBlock);
  });
}



/*==================================================================================
  FORM
==================================================================================*/

/* deselect radio button
/------------------------*/
function deselectRadioButtons(rootElement) {
  if(!rootElement) rootElement = document;
  if(!window.radioChecked) window.radioChecked = null;
  window.radioClick = function(e) {
    const obj = e.target;
    if(e.keyCode) return obj.checked = e.keyCode!=32;
    obj.checked = window.radioChecked != obj;
    window.radioChecked = obj.checked ? obj : null;
  }
  rootElement.querySelectorAll("input[type='radio']").forEach( radio => {
    radio.setAttribute("onclick", "radioClick(event)");
    radio.setAttribute("onkeyup", "radioClick(event)");
  });
}
deselectRadioButtons();



/*==================================================================================
  POP UP
==================================================================================*/

/* Create
/------------------------*/
function closePopUp(){
  // enable scrolling
  root.classList.remove('popup-noscroll');
  // hide and remove popup
  var popup = document.querySelector('.popup');
  popup.classList.add('closed');
  popup.remove();
}


/* Create
/------------------------*/
function loadPopUp(){
  const popup_close = '<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" viewBox="0 0 24.9 24.9" style="enable-background:new 0 0 24.9 24.9;" xml:space="preserve"><rect x="-3.7" y="10.9" transform="matrix(0.7071 -0.7071 0.7071 0.7071 -5.1549 12.4451)" fill="#fff" width="32.2" height="3"/><rect x="10.9" y="-3.7" transform="matrix(0.7071 -0.7071 0.7071 0.7071 -5.1549 12.4451)" fill="#fff" width="3" height="32.2"/></svg>',
        popup_content = '<div class="popup closed" data-content="img-popup"><div class="popup-container"><span class="close">' + popup_close + '</span><div class="popup-content"></div></div></div>';
  // inser pop up
  body.insertAdjacentHTML('beforeend', popup_content);
  // disable scrolling and show popup
  setTimeout(function() {
    var popup = document.querySelector('.popup');
    root.classList.add('popup-noscroll');
    popup.classList.remove('closed');
    // add closing event
    var closePopUpButtons = document.querySelectorAll('.popup > .popup-container > .close');
    if(closePopUpButtons.length > 0){
      Array.from(closePopUpButtons).forEach(function(closeButton) {
        closeButton.addEventListener('click', closePopUp);
        closeButton.addEventListener('keypress', closePopUp);
      });
    }
  }, 500);
}
